pkg/collect/all_logs.go

```diff
@@ -0,0 +1,201 @@
+	client, err := kubernetes.NewForConfig(c.ClientConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	output := NewResult()
+
+	ctx := context.Background()
+
+	// Set default logCollector Name if empty
+	logCollectorName := ""
+
+	if logsCollector.Name == "" {
+		logCollectorName = "allPodLogs"
+	} else {
+		logCollectorName = logsCollector.Name
+	}
+
+	// If Namespaces list is empty, get all the namespaces
+	if len(logsCollector.Namespaces) == 0 || logsCollector.Namespaces[0] == "*" {
+		_, namespaceList, namespaceErrors := getAllNamespaces(ctx, client)
+		if len(namespaceErrors) > 0 {
+			output.SaveResult(c.BundlePath, getAllLogsErrorsFileName(logsCollector), marshalErrors(namespaceErrors))
+		}
+		logsCollector.Namespaces = getNamespaceNames(namespaceList)
+	}
+
+	for _, namespace := range logsCollector.Namespaces {
+		pods, podsErrors := listPodsInNamespace(ctx, client, namespace, logsCollector.Selector)
+		if len(podsErrors) > 0 {
+			output.SaveResult(c.BundlePath, getAllLogsErrorsFileName(logsCollector), marshalErrors(podsErrors))
+		}
+
+		for _, pod := range pods {
+			if len(logsCollector.ContainerNames) == 0 {
+				// make a list of all the containers in the pod, so that we can get logs from all of them
+				containerNames := []string{}
+				for _, container := range pod.Spec.Containers {
+					containerNames = append(containerNames, container.Name)
+				}
+				for _, container := range pod.Spec.InitContainers {
+					containerNames = append(containerNames, container.Name)
+				}
+
+				for _, containerName := range containerNames {
+					if len(containerNames) == 1 {
+						containerName = "" // if there was only one container, use the old behavior of not including the container name in the path
+					}
+					podLogs, err := saveNamespacedPodLogs(ctx, c.BundlePath, client, pod, logCollectorName, containerName, namespace, logsCollector.Limits, false)
+					if err != nil {
+						key := fmt.Sprintf("%s/%s/%s-errors.json", namespace, logCollectorName, pod.Name)
+						if containerName != "" {
+							key = fmt.Sprintf("%s/%s/%s/%s-errors.json", namespace, logCollectorName, pod.Name, containerName)
+						}
+						output.SaveResult(c.BundlePath, key, marshalErrors([]string{err.Error()}))
+						if err != nil {
+							return nil, err
+						}
+						continue
+					}
+					for k, v := range podLogs {
+						output[k] = v
+					}
+				}
+			} else {
+				for _, container := range logsCollector.ContainerNames {
+					containerLogs, err := saveNamespacedPodLogs(ctx, c.BundlePath, client, pod, logCollectorName, container, namespace, logsCollector.Limits, false)
+					if err != nil {
+						key := fmt.Sprintf("%s/%s/%s/%s-errors.json", namespace, logCollectorName, pod.Name, container)
+						output.SaveResult(c.BundlePath, key, marshalErrors([]string{err.Error()}))
+						if err != nil {
+							return nil, err
+						}
+						continue
+					}
+					for k, v := range containerLogs {
+						output[k] = v
+					}
+				}
+			}
```

Comment:
This code block is very similar to the one above. Is there any chance we can dedupe these?

Reply:
Tried de-duping the looping block but the common func seems to have more than 10 arguments to make it work which does not look good to me. It looks clear this way that we are either looping through all the containers within a pod or specified containers.